# Package load

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns 
from matplotlib import pyplot as plt

# file load

In [2]:
_raw_read_count = pd.read_csv('221027_fat_total_readcount.txt', sep='\t', engine='python', encoding='cp949')
_raw_TPM_count = pd.read_csv('221027_fat_total_tpm.txt', sep='\t', engine='python', encoding='cp949')

display(_raw_read_count.head())
display(_raw_TPM_count.head())

FN42  FN45  FN47  FN48  FN50  FN51  FN52  FN53  FN54  FN56  ...  CFN02  \
0     0     0     0     0     0     0     4     0     0     0  ...      0   
1    30    91    36     5    38    20    55     0    37    22  ...      0   
2     0     0     0     0     0     0     0     0     0     0  ...      0   
3     0     0     0     0     0     0     0     0     0     0  ...      0   
4     0     0     0     0     0     0     0     0     0     0  ...      0   

   CFN03  CFN04  CFN05  CFN09  CFN11  RCFN01  RCFN06  RCFN07  RCFN12  
0      0      0      0      0      0       0       0       0       0  
1     28     54     86      0      0       0       0      20       0  
2      0      0      0      0      0       0       0       0       0  
3      0     13      0      0      0       0       0       0       0  
4      0      0      0      0      0       0       0       0       0  

[5 rows x 63 columns]

FN42      FN45      FN47      FN48      FN50      FN51      FN52  FN53  \
0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.085567   0.0   
1  0.727646  2.320532  0.843772  0.122902  0.798591  0.431733  1.531822   0.0   
2  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   
3  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   
4  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   

       FN54      FN56  ...  CFN02     CFN03     CFN04     CFN05  CFN09  CFN11  \
0  0.000000  0.000000  ...    0.0  0.000000  0.000000  0.000000    0.0    0.0   
1  1.046038  0.584521  ...    0.0  0.424949  1.173461  1.129727    0.0    0.0   
2  0.000000  0.000000  ...    0.0  0.000000  0.000000  0.000000    0.0    0.0   
3  0.000000  0.000000  ...    0.0  0.000000  0.433210  0.000000    0.0    0.0   
4  0.000000  0.000000  ...    0.0  0.000000  0.000000  0.000000    0.0    0.0   

   RCFN01  RCFN06    RCFN07  RCFN12  
0     0.0     0.0  0.000000     0.0  
1     0.0     0.0  0.302067     0.0  
2     0.0     0.0  0.000000     0.0  
3     0.0     0.0  0.000000     0.0  
4     0.0     0.0  0.000000     0.0  

[5 rows x 63 columns]

In [3]:
gtf = pd.read_csv('gencode.v38.genes.txt', sep='\t', engine='python', encoding='cp949')
display(gtf.head())

chromosome        type                      gene_id  \
0       chr1        gene  gene_id "ENSG00000223972.5"   
1       chr1  transcript  gene_id "ENSG00000223972.5"   
2       chr1        exon  gene_id "ENSG00000223972.5"   
3       chr1        exon  gene_id "ENSG00000223972.5"   
4       chr1        exon  gene_id "ENSG00000223972.5"   

                                         gene_type             gene_name  
0   gene_type "transcribed_unprocessed_pseudogene"   gene_name "DDX11L1"  
1   gene_type "transcribed_unprocessed_pseudogene"   gene_name "DDX11L1"  
2   gene_type "transcribed_unprocessed_pseudogene"   gene_name "DDX11L1"  
3   gene_type "transcribed_unprocessed_pseudogene"   gene_name "DDX11L1"  
4   gene_type "transcribed_unprocessed_pseudogene"   gene_name "DDX11L1"

In [4]:
name_des = pd.read_csv('name_des.txt', sep='\t', engine='python', encoding='cp949')
name_des.head()

Name  Description
0  ENSG00000223972.5      DDX11L1
1  ENSG00000227232.5       WASH7P
2  ENSG00000278267.1    MIR6859-1
3  ENSG00000243485.5  MIR1302-2HG
4  ENSG00000237613.2      FAM138A

In [5]:
_raw_TPM_count = pd.concat([name_des, _raw_TPM_count], axis = 1)
raw_TPM_count = _raw_TPM_count.drop(columns=['FP47', 'FN47', 'FP53', 'FN53'])
raw_TPM_count

Name  Description        FN42       FN45        FN48  \
0      ENSG00000223972.5      DDX11L1    0.000000   0.000000    0.000000   
1      ENSG00000227232.5       WASH7P    0.727646   2.320532    0.122902   
2      ENSG00000278267.1    MIR6859-1    0.000000   0.000000    0.000000   
3      ENSG00000243485.5  MIR1302-2HG    0.000000   0.000000    0.000000   
4      ENSG00000237613.2      FAM138A    0.000000   0.000000    0.000000   
...                  ...          ...         ...        ...         ...   
58199  ENSG00000198695.2       MT-ND6    9.571104   4.386254    6.715108   
58200  ENSG00000210194.1        MT-TE    0.000000   0.000000    0.000000   
58201  ENSG00000198727.2       MT-CYB  114.331562  71.824716  156.205221   
58202  ENSG00000210195.2        MT-TT    0.000000   0.000000    0.000000   
58203  ENSG00000210196.2        MT-TP    0.000000   2.490034    0.000000   

             FN50       FN51        FN52       FN54        FN56  ...  \
0        0.000000   0.000000    0.085567   0.000000    0.000000  ...   
1        0.798591   0.431733    1.531822   1.046038    0.584521  ...   
2        0.000000   0.000000    0.000000   0.000000    0.000000  ...   
3        0.000000   0.000000    0.000000   0.000000    0.000000  ...   
4        0.000000   0.000000    0.000000   0.000000    0.000000  ...   
...           ...        ...         ...        ...         ...  ...   
58199    6.325959   3.767673   10.426682  10.798452    8.064875  ...   
58200    0.000000   0.000000    0.000000   0.000000    0.000000  ...   
58201  139.616588  67.986948  127.491726  96.674069  111.665116  ...   
58202    0.000000   0.000000    0.000000   0.000000    0.000000  ...   
58203    0.000000   0.000000    0.000000   1.104244    6.745437  ...   

           CFN02      CFN03       CFN04      CFN05      CFN09       CFN11  \
0       0.000000   0.000000    0.000000   0.000000   0.000000    0.000000   
1       0.000000   0.424949    1.173461   1.129727   0.000000    0.000000   
2       0.000000   0.000000    0.000000   0.000000   0.000000    0.000000   
3       0.000000   0.000000    0.433210   0.000000   0.000000    0.000000   
4       0.000000   0.000000    0.000000   0.000000   0.000000    0.000000   
...          ...        ...         ...        ...        ...         ...   
58199   8.253482   6.257558   13.632186   7.044494   7.455114    6.231438   
58200   0.000000   0.000000    0.000000   0.000000   0.000000    0.000000   
58201  90.410917  48.576209  111.007692  97.744498  92.120733  149.768920   
58202   0.000000   0.000000    0.000000   0.000000   0.000000    0.000000   
58203   0.000000   0.000000    0.000000   0.000000   0.000000    0.000000   

           RCFN01     RCFN06     RCFN07      RCFN12  
0        0.000000   0.000000   0.000000    0.000000  
1        0.000000   0.000000   0.302067    0.000000  
2        0.000000   0.000000   0.000000    0.000000  
3        0.000000   0.000000   0.000000    0.000000  
4        0.000000   0.000000   0.000000    0.000000  
...           ...        ...        ...         ...  
58199   10.106898  12.130103   7.488040   11.380422  
58200    0.000000   0.000000   0.000000    0.000000  
58201  155.498197  65.835316  78.752395  157.766099  
58202    0.000000   0.000000   0.000000    0.000000  
58203    0.000000   0.000000   0.000000    0.000000  

[58204 rows x 61 columns]

In [6]:
_raw_read_count = pd.concat([name_des, _raw_read_count], axis = 1)
raw_read_count = _raw_read_count.drop(columns=['FP47', 'FN47', 'FP53', 'FN53'])
raw_read_count

Name  Description  FN42  FN45  FN48  FN50  FN51  FN52  \
0      ENSG00000223972.5      DDX11L1     0     0     0     0     0     4   
1      ENSG00000227232.5       WASH7P    30    91     5    38    20    55   
2      ENSG00000278267.1    MIR6859-1     0     0     0     0     0     0   
3      ENSG00000243485.5  MIR1302-2HG     0     0     0     0     0     0   
4      ENSG00000237613.2      FAM138A     0     0     0     0     0     0   
...                  ...          ...   ...   ...   ...   ...   ...   ...   
58199  ENSG00000198695.2       MT-ND6   156    68   108   119    69   148   
58200  ENSG00000210194.1        MT-TE     0     0     0     0     0     0   
58201  ENSG00000198727.2       MT-CYB  4050  2420  5460  5708  2706  3933   
58202  ENSG00000210195.2        MT-TT     0     0     0     0     0     0   
58203  ENSG00000210196.2        MT-TP     0     5     0     0     0     0   

       FN54  FN56  ...  CFN02  CFN03  CFN04  CFN05  CFN09  CFN11  RCFN01  \
0         0     0  ...      0      0      0      0      0      0       0   
1        37    22  ...      0     28     54     86      0      0       0   
2         0     0  ...      0      0      0      0      0      0       0   
3         0     0  ...      0      0     13      0      0      0       0   
4         0     0  ...      0      0      0      0      0      0       0   
...     ...   ...  ...    ...    ...    ...    ...    ...    ...     ...   
58199   151   120  ...    192    163    248    212    228    115     176   
58200     0     0  ...      0      0      0      0      0      0       0   
58201  2938  3611  ...   4571   2750   4389   6393   6123   6007    5885   
58202     0     0  ...      0      0      0      0      0      0       0   
58203     2    13  ...      0      0      0      0      0      0       0   

       RCFN06  RCFN07  RCFN12  
0           0       0       0  
1           0      20       0  
2           0       0       0  
3           0       0       0  
4           0       0       0  
...       ...     ...     ...  
58199     274     196     264  
58200       0       0       0  
58201    3232    4480    7954  
58202       0       0       0  
58203       0       0       0  

[58204 rows x 61 columns]

# Gene QC

## raw data processing

In [7]:
len(raw_read_count.iloc[0][2:][raw_read_count.iloc[0][2:] >= 6]) > 30

False

In [8]:
idx_1 = []
for i in range(len(raw_read_count)):
    if len(raw_read_count.iloc[i][2:][raw_read_count.iloc[i][2:] >= 6]) >= 30:
        idx_1.append(i)

In [9]:
_read_count = raw_read_count.loc[idx_1]
display(_read_count.head())
print(len(_read_count))

Name    Description  FN42  FN45  FN48  FN50  FN51  FN52  \
1    ENSG00000227232.5         WASH7P    30    91     5    38    20    55   
13   ENSG00000241860.7  RP11-34P13.13    27    27    55    46    34    16   
17   ENSG00000279457.4         WASH9P    19    42    20    88   110    43   
19  ENSG00000228463.10     AP006222.1    56    90     0    44    60    17   
23  ENSG00000237094.12   RP4-669L17.4    63    44    51    29    62    49   

    FN54  FN56  ...  CFN02  CFN03  CFN04  CFN05  CFN09  CFN11  RCFN01  RCFN06  \
1     37    22  ...      0     28     54     86      0      0       0       0   
13    44    62  ...     27      5     40      2      0     58     156       0   
17    16    14  ...    136    110     42     54    132     59     223       0   
19    70    58  ...     70      0     18     24      0     16       0       0   
23    47    75  ...      3     37     22      1     43     62       0      33   

    RCFN07  RCFN12  
1       20       0  
13      47      94  
17       0      60  
19     181       6  
23       1      43  

[5 rows x 61 columns]

20871


In [10]:
idx_2 = []
for i in range(len(raw_TPM_count)):
    if len(raw_TPM_count.iloc[i][2:][raw_TPM_count.iloc[i][2:] >= 0.1]) >= 30:
        idx_2.append(i)

In [11]:
_TPM_count = raw_TPM_count.loc[idx_2]
display(_TPM_count.head())
print(len(_TPM_count))

Name   Description      FN42      FN45      FN48      FN50  \
1    ENSG00000227232.5        WASH7P  0.727646  2.320532  0.122902  0.798591   
17   ENSG00000279457.4        WASH9P  0.438081  1.018116  0.467328  1.758025   
23  ENSG00000237094.12  RP4-669L17.4  0.347535  0.255187  0.285115  0.138611   
32   ENSG00000225630.1      MTND2P28  1.789469  0.648745  0.375206  0.801972   
33   ENSG00000237973.1      MTCO1P12  3.547817  2.085855  5.109321  2.588628   

         FN51      FN52      FN54      FN56  ...     CFN02     CFN03  \
1    0.431733  1.531822  1.046038  0.584521  ...  0.000000  0.424949   
17   2.257250  1.138455  0.429999  0.353596  ...  2.197039  1.586985   
23   0.304394  0.310385  0.302206  0.453209  ...  0.011595  0.127714   
32   0.329507  0.921121  1.941946  0.000000  ...  1.037615  1.158317   
33  35.293741  4.154285  2.775012  2.761343  ...  3.401571  1.460486   

       CFN04     CFN05     CFN09     CFN11    RCFN01    RCFN06    RCFN07  \
1   1.173461  1.129727  0.000000  0.000000  0.000000  0.000000  0.302067   
17  0.867613  0.674327  1.622020  1.201449  4.812525  0.000000  0.000000   
23  0.108732  0.002988  0.126418  0.302067  0.000000  0.131356  0.003435   
32  0.663413  0.200518  1.677175  1.580437  1.039601  1.870045  0.000000   
33  3.094965  3.501660  3.296443  4.597010  4.216409  3.705244  3.110327   

      RCFN12  
1   0.000000  
17  0.972005  
23  0.166665  
32  1.517439  
33  3.968716  

[5 rows x 61 columns]

20765


In [12]:
_TPM_count['gene_name'] = _TPM_count['Name'] + "_" + _TPM_count['Description']
_TPM_count.head()

Name   Description      FN42      FN45      FN48      FN50  \
1    ENSG00000227232.5        WASH7P  0.727646  2.320532  0.122902  0.798591   
17   ENSG00000279457.4        WASH9P  0.438081  1.018116  0.467328  1.758025   
23  ENSG00000237094.12  RP4-669L17.4  0.347535  0.255187  0.285115  0.138611   
32   ENSG00000225630.1      MTND2P28  1.789469  0.648745  0.375206  0.801972   
33   ENSG00000237973.1      MTCO1P12  3.547817  2.085855  5.109321  2.588628   

         FN51      FN52      FN54      FN56  ...     CFN03     CFN04  \
1    0.431733  1.531822  1.046038  0.584521  ...  0.424949  1.173461   
17   2.257250  1.138455  0.429999  0.353596  ...  1.586985  0.867613   
23   0.304394  0.310385  0.302206  0.453209  ...  0.127714  0.108732   
32   0.329507  0.921121  1.941946  0.000000  ...  1.158317  0.663413   
33  35.293741  4.154285  2.775012  2.761343  ...  1.460486  3.094965   

       CFN05     CFN09     CFN11    RCFN01    RCFN06    RCFN07    RCFN12  \
1   1.129727  0.000000  0.000000  0.000000  0.000000  0.302067  0.000000   
17  0.674327  1.622020  1.201449  4.812525  0.000000  0.000000  0.972005   
23  0.002988  0.126418  0.302067  0.000000  0.131356  0.003435  0.166665   
32  0.200518  1.677175  1.580437  1.039601  1.870045  0.000000  1.517439   
33  3.501660  3.296443  4.597010  4.216409  3.705244  3.110327  3.968716   

                          gene_name  
1          ENSG00000227232.5_WASH7P  
17         ENSG00000279457.4_WASH9P  
23  ENSG00000237094.12_RP4-669L17.4  
32       ENSG00000225630.1_MTND2P28  
33       ENSG00000237973.1_MTCO1P12  

[5 rows x 62 columns]

In [13]:
_read_count['gene_name'] = _read_count['Name'] + "_" + _read_count['Description']
_read_count.head()

Name    Description  FN42  FN45  FN48  FN50  FN51  FN52  \
1    ENSG00000227232.5         WASH7P    30    91     5    38    20    55   
13   ENSG00000241860.7  RP11-34P13.13    27    27    55    46    34    16   
17   ENSG00000279457.4         WASH9P    19    42    20    88   110    43   
19  ENSG00000228463.10     AP006222.1    56    90     0    44    60    17   
23  ENSG00000237094.12   RP4-669L17.4    63    44    51    29    62    49   

    FN54  FN56  ...  CFN03  CFN04  CFN05  CFN09  CFN11  RCFN01  RCFN06  \
1     37    22  ...     28     54     86      0      0       0       0   
13    44    62  ...      5     40      2      0     58     156       0   
17    16    14  ...    110     42     54    132     59     223       0   
19    70    58  ...      0     18     24      0     16       0       0   
23    47    75  ...     37     22      1     43     62       0      33   

    RCFN07  RCFN12                        gene_name  
1       20       0         ENSG00000227232.5_WASH7P  
13      47      94  ENSG00000241860.7_RP11-34P13.13  
17       0      60         ENSG00000279457.4_WASH9P  
19     181       6    ENSG00000228463.10_AP006222.1  
23       1      43  ENSG00000237094.12_RP4-669L17.4  

[5 rows x 62 columns]

In [14]:
read_count_s = set(_read_count['gene_name'])
print(len(read_count_s))

20871


In [15]:
TPM_coun_s = set(_TPM_count['gene_name'])
print(len(TPM_coun_s))

20765


In [16]:
we_gene_list = read_count_s & TPM_coun_s
len(we_gene_list)

20109

In [17]:
_read_count[_read_count['gene_name'].isin(list(we_gene_list))]

Name   Description  FN42  FN45  FN48  FN50  FN51  FN52  \
1       ENSG00000227232.5        WASH7P    30    91     5    38    20    55   
17      ENSG00000279457.4        WASH9P    19    42    20    88   110    43   
23     ENSG00000237094.12  RP4-669L17.4    63    44    51    29    62    49   
32      ENSG00000225630.1      MTND2P28    58    20    12    30    12    26   
33      ENSG00000237973.1      MTCO1P12   152    85   216   128  1699   155   
...                   ...           ...   ...   ...   ...   ...   ...   ...   
58193   ENSG00000212907.2       MT-ND4L   475   273   412   556   266   396   
58194   ENSG00000198886.2        MT-ND4  2420  1268  2198  1929  1418  1781   
58198   ENSG00000198786.2        MT-ND5  1326  1232  1994  2134  1356  1452   
58199   ENSG00000198695.2        MT-ND6   156    68   108   119    69   148   
58201   ENSG00000198727.2        MT-CYB  4050  2420  5460  5708  2706  3933   

       FN54  FN56  ...  CFN03  CFN04  CFN05  CFN09  CFN11  RCFN01  RCFN06  \
1        37    22  ...     28     54     86      0      0       0       0   
17       16    14  ...    110     42     54    132     59     223       0   
23       47    75  ...     37     22      1     43     62       0      33   
32       54     0  ...     60     24     12    102     58      36      84   
33      102   108  ...    100    148    277    265    223     193     220   
...     ...   ...  ...    ...    ...    ...    ...    ...     ...     ...   
58193   398   492  ...    388    306    777   1214    544     596     310   
58194  1777  1664  ...   1342   2152   2684   3016   2507    2922    2181   
58198  1428  1420  ...   1614   2412   1989   3660   1967    3438    2310   
58199   151   120  ...    163    248    212    228    115     176     274   
58201  2938  3611  ...   2750   4389   6393   6123   6007    5885    3232   

       RCFN07  RCFN12                        gene_name  
1          20       0         ENSG00000227232.5_WASH7P  
17          0      60         ENSG00000279457.4_WASH9P  
23          1      43  ENSG00000237094.12_RP4-669L17.4  
32          0      70       ENSG00000225630.1_MTND2P28  
33        214     242       ENSG00000237973.1_MTCO1P12  
...       ...     ...                              ...  
58193     560    1216        ENSG00000212907.2_MT-ND4L  
58194    2247    3114         ENSG00000198886.2_MT-ND4  
58198    2560    2795         ENSG00000198786.2_MT-ND5  
58199     196     264         ENSG00000198695.2_MT-ND6  
58201    4480    7954         ENSG00000198727.2_MT-CYB  

[20109 rows x 62 columns]

In [18]:
_TPM_count[_TPM_count['gene_name'].isin(list(we_gene_list))]

Name   Description        FN42       FN45        FN48  \
1       ENSG00000227232.5        WASH7P    0.727646   2.320532    0.122902   
17      ENSG00000279457.4        WASH9P    0.438081   1.018116    0.467328   
23     ENSG00000237094.12  RP4-669L17.4    0.347535   0.255187    0.285115   
32      ENSG00000225630.1      MTND2P28    1.789469   0.648745    0.375206   
33      ENSG00000237973.1      MTCO1P12    3.547817   2.085855    5.109321   
...                   ...           ...         ...        ...         ...   
58193   ENSG00000212907.2       MT-ND4L   52.758491  31.879305   46.375410   
58194   ENSG00000198886.2        MT-ND4   56.855786  31.320306   52.333380   
58198   ENSG00000198786.2        MT-ND5   23.571221  23.024845   35.921586   
58199   ENSG00000198695.2        MT-ND6    9.571104   4.386254    6.715108   
58201   ENSG00000198727.2        MT-CYB  114.331562  71.824716  156.205221   

             FN50       FN51        FN52       FN54        FN56  ...  \
1        0.798591   0.431733    1.531822   1.046038    0.584521  ...   
17       1.758025   2.257250    1.138455   0.429999    0.353596  ...   
23       0.138611   0.304394    0.310385   0.302206    0.453209  ...   
32       0.801972   0.329507    0.921121   1.941946    0.000000  ...   
33       2.588628  35.293741    4.154285   2.775012    2.761343  ...   
...           ...        ...         ...        ...         ...  ...   
58193   53.507603  26.294628   50.505757  51.526296   59.860837  ...   
58194   39.267520  29.649828   48.047430  48.662434   42.824425  ...   
58198   32.868124  21.452845   29.638220  29.587881   27.650681  ...   
58199    6.325959   3.767673   10.426682  10.798452    8.064875  ...   
58201  139.616588  67.986948  127.491726  96.674069  111.665116  ...   

           CFN03       CFN04      CFN05      CFN09       CFN11      RCFN01  \
1       0.424949    1.173461   1.129727   0.000000    0.000000    0.000000   
17      1.586985    0.867613   0.674327   1.622020    1.201449    4.812525   
23      0.127714    0.108732   0.002988   0.126418    0.302067    0.000000   
32      1.158317    0.663413   0.200518   1.677175    1.580437    1.039601   
33      1.460486    3.094965   3.501660   3.296443    4.597010    4.216409   
...          ...         ...        ...        ...         ...         ...   
58193  26.965613   30.450650  46.740815  71.862027   53.364275   61.960198   
58194  19.728389   45.297888  34.152171  37.763536   52.019550   64.255054   
58198  17.952375   38.414240  19.149165  34.673818   30.881321   57.202132   
58199   6.257558   13.632186   7.044494   7.455114    6.231438   10.106898   
58201  48.576209  111.007692  97.744498  92.120733  149.768920  155.498197   

          RCFN06     RCFN07      RCFN12                        gene_name  
1       0.000000   0.302067    0.000000         ENSG00000227232.5_WASH7P  
17      0.000000   0.000000    0.972005         ENSG00000279457.4_WASH9P  
23      0.131356   0.003435    0.166665  ENSG00000237094.12_RP4-669L17.4  
32      1.870045   0.000000    1.517439       ENSG00000225630.1_MTND2P28  
33      3.705244   3.110327    3.968716       ENSG00000237973.1_MTCO1P12  
...          ...        ...         ...                              ...  
58193  24.844881  38.731243   94.896309        ENSG00000212907.2_MT-ND4L  
58194  36.973574  32.872824   51.403761         ENSG00000198886.2_MT-ND4  
58198  29.629683  28.336955   34.908996         ENSG00000198786.2_MT-ND5  
58199  12.130103   7.488040   11.380422         ENSG00000198695.2_MT-ND6  
58201  65.835316  78.752395  157.766099         ENSG00000198727.2_MT-CYB  

[20109 rows x 62 columns]

## GTF

### 파일 정리

In [19]:
v3 = gtf['gene_id'].str.split('"')
v3_1 = gtf['gene_type'].str.split('"')
v3_2 = gtf['gene_name'].str.split('"')
gtf['gene_id'] = v3.str.get(1)
gtf['gene_type'] = v3_1.str.get(1)
gtf['gene_name'] = v3_2.str.get(1)

In [20]:
gtf

chromosome        type            gene_id  \
0            chr1        gene  ENSG00000223972.5   
1            chr1  transcript  ENSG00000223972.5   
2            chr1        exon  ENSG00000223972.5   
3            chr1        exon  ENSG00000223972.5   
4            chr1        exon  ENSG00000223972.5   
...           ...         ...                ...   
465022       chrM  transcript  ENSG00000210195.2   
465023       chrM        exon  ENSG00000210195.2   
465024       chrM        gene  ENSG00000210196.2   
465025       chrM  transcript  ENSG00000210196.2   
465026       chrM        exon  ENSG00000210196.2   

                                 gene_type gene_name  
0       transcribed_unprocessed_pseudogene   DDX11L1  
1       transcribed_unprocessed_pseudogene   DDX11L1  
2       transcribed_unprocessed_pseudogene   DDX11L1  
3       transcribed_unprocessed_pseudogene   DDX11L1  
4       transcribed_unprocessed_pseudogene   DDX11L1  
...                                    ...       ...  
465022                             Mt_tRNA     MT-TT  
465023                             Mt_tRNA     MT-TT  
465024                             Mt_tRNA     MT-TP  
465025                             Mt_tRNA     MT-TP  
465026                             Mt_tRNA     MT-TP  

[465027 rows x 5 columns]

In [21]:
gtf['chromosome'].value_counts()

chr1     43790
chr2     34893
chr3     26791
chr11    25080
chr12    24740
chr17    24459
chr19    23125
chr6     22969
chr7     22930
chr5     22080
chr4     20058
chr16    19810
chr10    18519
chr8     18425
chr9     17968
chr15    17504
chrX     16297
chr14    15783
chr20    10982
chr22    10039
chr13     9915
chr18     9007
chr21     6184
chrY      3568
chrM       111
Name: chromosome, dtype: int64

chrM : 미토콘드리아 염색체

### chromosome X, Y 제거 

In [22]:
gtf['type'].value_counts()

exon          348619
gene           58204
transcript     58204
Name: type, dtype: int64

In [23]:
gtfxy = gtf[(gtf['chromosome'] != 'chrX') & (gtf['chromosome'] != 'chrY')]
gtfxy

chromosome        type            gene_id  \
0            chr1        gene  ENSG00000223972.5   
1            chr1  transcript  ENSG00000223972.5   
2            chr1        exon  ENSG00000223972.5   
3            chr1        exon  ENSG00000223972.5   
4            chr1        exon  ENSG00000223972.5   
...           ...         ...                ...   
465022       chrM  transcript  ENSG00000210195.2   
465023       chrM        exon  ENSG00000210195.2   
465024       chrM        gene  ENSG00000210196.2   
465025       chrM  transcript  ENSG00000210196.2   
465026       chrM        exon  ENSG00000210196.2   

                                 gene_type gene_name  
0       transcribed_unprocessed_pseudogene   DDX11L1  
1       transcribed_unprocessed_pseudogene   DDX11L1  
2       transcribed_unprocessed_pseudogene   DDX11L1  
3       transcribed_unprocessed_pseudogene   DDX11L1  
4       transcribed_unprocessed_pseudogene   DDX11L1  
...                                    ...       ...  
465022                             Mt_tRNA     MT-TT  
465023                             Mt_tRNA     MT-TT  
465024                             Mt_tRNA     MT-TP  
465025                             Mt_tRNA     MT-TP  
465026                             Mt_tRNA     MT-TP  

[445162 rows x 5 columns]

### Exon, gene, transcript 중 Exon, transcript 제거 -> gene만 남기기

In [24]:
gtf_gene = gtfxy[gtfxy['type'] == 'gene']
gtf_gene

chromosome  type            gene_id  \
0            chr1  gene  ENSG00000223972.5   
6            chr1  gene  ENSG00000227232.5   
19           chr1  gene  ENSG00000278267.1   
22           chr1  gene  ENSG00000243485.5   
28           chr1  gene  ENSG00000237613.2   
...           ...   ...                ...   
465012       chrM  gene  ENSG00000198695.2   
465015       chrM  gene  ENSG00000210194.1   
465018       chrM  gene  ENSG00000198727.2   
465021       chrM  gene  ENSG00000210195.2   
465024       chrM  gene  ENSG00000210196.2   

                                 gene_type    gene_name  
0       transcribed_unprocessed_pseudogene      DDX11L1  
6                   unprocessed_pseudogene       WASH7P  
19                                   miRNA    MIR6859-1  
22                                  lncRNA  MIR1302-2HG  
28                                  lncRNA      FAM138A  
...                                    ...          ...  
465012                      protein_coding       MT-ND6  
465015                             Mt_tRNA        MT-TE  
465018                      protein_coding       MT-CYB  
465021                             Mt_tRNA        MT-TT  
465024                             Mt_tRNA        MT-TP  

[55290 rows x 5 columns]

### Protein coding, lncRNA 남기기

In [25]:
len(gtfxy['gene_type'].value_counts())

40

protein_coding                        274012 <br>
lncRNA                                 92341 <br>
processed_pseudogene                   29138 <br>
unprocessed_pseudogene                 11077<br>
transcribed_unprocessed_pseudogene      8849<br>
misc_RNA                                5705<br>
snRNA                                   5248<br>
miRNA                                   4351<br>
TEC                                     3205<br>
snoRNA                                  2251<br>
transcribed_processed_pseudogene        2188<br>
transcribed_unitary_pseudogene          1822<br>
rRNA_pseudogene                         1375<br>
IG_V_pseudogene                          661<br>
IG_V_gene                                578<br>
unitary_pseudogene                       471<br>
TR_V_gene                                423<br>
polymorphic_pseudogene                   334<br>
TR_J_gene                                237<br>
rRNA                                     135<br>
TR_V_pseudogene                          123<br>
scaRNA                                   111<br>
IG_D_gene                                111<br>
IG_C_gene                                 91<br>
Mt_tRNA                                   66<br>
IG_J_gene                                 54<br>
pseudogene                                48<br>
TR_C_gene                                 34<br>
IG_C_pseudogene                           33<br>
ribozyme                                  15<br>
TR_D_gene                                 12<br>
TR_J_pseudogene                           12<br>
sRNA                                      12<br>
translated_unprocessed_pseudogene          9<br>
IG_J_pseudogene                            9<br>
translated_processed_pseudogene            6<br>
Mt_rRNA                                    6<br>
vault_RNA                                  3<br>
IG_pseudogene                              3<br>
scRNA                                      3<br>

In [26]:
gtf_type = gtf_gene[(gtf_gene['gene_type'] == 'protein_coding') | (gtf_gene['gene_type'] == 'lncRNA')]
gtf_type

chromosome  type            gene_id       gene_type      gene_name
22           chr1  gene  ENSG00000243485.5          lncRNA    MIR1302-2HG
28           chr1  gene  ENSG00000237613.2          lncRNA        FAM138A
41           chr1  gene  ENSG00000186092.7  protein_coding          OR4F5
46           chr1  gene  ENSG00000238009.6          lncRNA   RP11-34P13.7
66           chr1  gene  ENSG00000239906.1          lncRNA  RP11-34P13.14
...           ...   ...                ...             ...            ...
464994       chrM  gene  ENSG00000212907.2  protein_coding        MT-ND4L
464997       chrM  gene  ENSG00000198886.2  protein_coding         MT-ND4
465009       chrM  gene  ENSG00000198786.2  protein_coding         MT-ND5
465012       chrM  gene  ENSG00000198695.2  protein_coding         MT-ND6
465018       chrM  gene  ENSG00000198727.2  protein_coding         MT-CYB

[34183 rows x 5 columns]

### 우리 데이터 gene 이름과 동일하게 만들기

In [27]:
gtf_type['gene'] = gtf_type['gene_id'] + '_' + gtf_type['gene_name']
gtf_type

C:\Users\user\AppData\Local\Temp\ipykernel_1740\2129002817.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  gtf_type['gene'] = gtf_type['gene_id'] + '_' + gtf_type['gene_name']


chromosome  type            gene_id       gene_type      gene_name  \
22           chr1  gene  ENSG00000243485.5          lncRNA    MIR1302-2HG   
28           chr1  gene  ENSG00000237613.2          lncRNA        FAM138A   
41           chr1  gene  ENSG00000186092.7  protein_coding          OR4F5   
46           chr1  gene  ENSG00000238009.6          lncRNA   RP11-34P13.7   
66           chr1  gene  ENSG00000239906.1          lncRNA  RP11-34P13.14   
...           ...   ...                ...             ...            ...   
464994       chrM  gene  ENSG00000212907.2  protein_coding        MT-ND4L   
464997       chrM  gene  ENSG00000198886.2  protein_coding         MT-ND4   
465009       chrM  gene  ENSG00000198786.2  protein_coding         MT-ND5   
465012       chrM  gene  ENSG00000198695.2  protein_coding         MT-ND6   
465018       chrM  gene  ENSG00000198727.2  protein_coding         MT-CYB   

                                   gene  
22        ENSG00000243485.5_MIR1302-2HG  
28            ENSG00000237613.2_FAM138A  
41              ENSG00000186092.7_OR4F5  
46       ENSG00000238009.6_RP11-34P13.7  
66      ENSG00000239906.1_RP11-34P13.14  
...                                 ...  
464994        ENSG00000212907.2_MT-ND4L  
464997         ENSG00000198886.2_MT-ND4  
465009         ENSG00000198786.2_MT-ND5  
465012         ENSG00000198695.2_MT-ND6  
465018         ENSG00000198727.2_MT-CYB  

[34183 rows x 6 columns]

### 중복확인

In [28]:
display(gtf_type[gtf_type['gene'].duplicated()])
print(len(gtf_type[gtf_type['gene'].duplicated()]))

Empty DataFrame
Columns: [chromosome, type, gene_id, gene_type, gene_name, gene]
Index: []

0


In [29]:
len(gtf_type[gtf_type['chromosome'] == 'chrM'])

13

### gtf_gene_list 뽑아오기

In [30]:
gtf_gene_list = set(gtf_type['gene'])
len(gtf_gene_list)

34183

## final gene & GTF 

In [31]:
## 우리 데이터 gene 이름 
print(len(we_gene_list))
print(len(gtf_gene_list))

20109
34183


In [32]:
final_gene_list = we_gene_list & gtf_gene_list
print(len(final_gene_list))

17268


In [33]:
list(final_gene_list)

['ENSG00000182670.13_TTC3',
 'ENSG00000164035.10_EMCN',
 'ENSG00000206077.14_ZDHHC11B',
 'ENSG00000287825.1_RP3-475N16.9',
 'ENSG00000225479.1_PLCB1-IT1',
 'ENSG00000153885.15_KCTD15',
 'ENSG00000147586.10_MRPS28',
 'ENSG00000154710.18_RABGEF1',
 'ENSG00000115317.12_HTRA2',
 'ENSG00000105647.19_PIK3R2',
 'ENSG00000272288.6_ILRUN-AS1',
 'ENSG00000259935.1_RP11-519C12.1',
 'ENSG00000156486.8_KCNS2',
 'ENSG00000187193.9_MT1X',
 'ENSG00000135912.11_TTLL4',
 'ENSG00000157551.19_KCNJ15',
 'ENSG00000213471.10_TTLL13P',
 'ENSG00000274211.6_SOCS7',
 'ENSG00000229666.1_MAST4-AS1',
 'ENSG00000274618.2_H4C6',
 'ENSG00000136636.13_KCTD3',
 'ENSG00000118620.13_ZNF430',
 'ENSG00000189337.17_KAZN',
 'ENSG00000139174.12_PRICKLE1',
 'ENSG00000164114.19_MAP9',
 'ENSG00000112624.12_BICRAL',
 'ENSG00000108107.15_RPL28',
 'ENSG00000170234.13_PWWP2A',
 'ENSG00000179598.6_PLD6',
 'ENSG00000005961.19_ITGA2B',
 'ENSG00000259865.1_RP11-488L18.10',
 'ENSG00000197122.12_SRC',
 'ENSG00000272836.1_RP3-402G11.27',
 '

### 최종 read_count, TPM_count 추출

In [34]:
read_count = _read_count[_read_count['gene_name'].isin(list(final_gene_list))]
TPM_count = _TPM_count[_TPM_count['gene_name'].isin(list(final_gene_list))]

### 샘플정보

In [35]:
_sample_file = pd.read_csv('Sample information_59_CrFCDCon.txt', sep='\t', engine='python', encoding='cp949')
_sample_file.index = list(_sample_file['Sample'])
sample_file = _sample_file.sort_index()
sample_file = sample_file[['Sex', 'Type']]
sample_file

Sex Type
CFN02    F  Con
CFN03    F  Con
CFN04    F  Con
CFN05    M  Con
CFN09    F  Con
CFN11    M  Con
FN42     M   CD
FN45     F   CD
FN48     M   CD
FN50     F   CD
FN51     M   CD
FN52     F   CD
FN54     M   CD
FN56     F   CD
FN57     F   CD
FN58     M   CD
FN60     M   CD
FN61     F   CD
FN62     M   CD
FN63     M   CD
FN65     M   CD
FN67     M   CD
FN68     M   CD
FN69     M   CD
FN71     M   CD
FN72     M   CD
FN74     M   CD
FN77     F   CD
FN82     M   CD
FP42     M  CrF
FP45     F  CrF
FP48     M  CrF
FP50     F  CrF
FP51     M  CrF
FP52     F  CrF
FP54     M  CrF
FP56     F  CrF
FP57     F  CrF
FP58     M  CrF
FP60     M  CrF
FP61     F  CrF
FP62     M  CrF
FP63     M  CrF
FP65     M  CrF
FP67     M  CrF
FP68     M  CrF
FP69     M  CrF
FP71     M  CrF
FP72     M  CrF
FP74     M  CrF
FP77     F  CrF
FP82     M  CrF
RCFN01   M  Con
RCFN06   M  Con
RCFN07   F  Con
RCFN12   F  Con
UC100    F  Con
UC92     F  Con
UC94     F  Con

- Number : 환자 고유 정보
- bulk : 우리는 fat인데 유코자랑 bulk가 곂치는 안 곂치는지에 대한 유무(1: 곂침, 0: 안 곂침), 무슨말인지 모르겠지만 안 쓰는 컬럼이라 깊게 안 물어봄

# PyDESeq2

In [36]:
import pydeseq2
pydeseq2.__version__

'0.4.0'

In [37]:
import os
import pickle as pkl

import pandas as pd

from pydeseq2.dds import DeseqDataSet
from pydeseq2.ds import DeseqStats

# Replace this with the path to directory where you would like results to be saved
OUTPUT_PATH = "../output_files"
os.makedirs(OUTPUT_PATH, exist_ok=True)  # Create path if it doesn't exist

In [38]:
_read_count_dds = read_count

In [39]:
_read_count_dds.index = list(_read_count_dds['gene_name'])
read_count_dds = _read_count_dds.T[2:-1]
read_count_dds = read_count_dds.sort_index()
read_count_dds

ENSG00000237491.10_LINC01409 ENSG00000177757.2_FAM87B  \
CFN02                           116                      110   
CFN03                           137                      178   
CFN04                           225                      175   
CFN05                           179                       24   
CFN09                           362                      202   
CFN11                           477                      268   
FN42                             93                      104   
FN45                            187                      111   
FN48                             89                       48   
FN50                            236                       70   
FN51                            231                       35   
FN52                            133                       87   
FN54                            152                       41   
FN56                             94                       68   
FN57                             92                       32   
FN58                             90                      122   
FN60                            148                       72   
FN61                            230                        7   
FN62                            125                      122   
FN63                            106                       22   
FN65                             78                       70   
FN67                            180                       58   
FN68                             93                       10   
FN69                            173                      155   
FN71                             39                       72   
FN72                             63                       54   
FN74                            262                      173   
FN77                            110                      143   
FN82                            205                      118   
FP42                            147                       75   
FP45                            211                       60   
FP48                            283                      109   
FP50                            276                      125   
FP51                            239                       64   
FP52                            235                       26   
FP54                            298                       29   
FP56                            160                       34   
FP57                            216                       26   
FP58                            162                       44   
FP60                             55                       25   
FP61                             67                      170   
FP62                              0                       20   
FP63                            137                      140   
FP65                            421                       82   
FP67                            238                       32   
FP68                            230                       18   
FP69                            195                       83   
FP71                            178                       48   
FP72                             97                       28   
FP74                            254                       65   
FP77                            163                       42   
FP82                            236                       72   
RCFN01                          348                      158   
RCFN06                          260                        0   
RCFN07                           96                      104   
RCFN12                           54                       98   
UC100                           145                        0   
UC92                             86                       12   
UC94                             65                       48   

       ENSG00000228794.10_LINC01128 ENSG00000187634.13_SAMD11  \
CFN02                          1358                       228   
CFN03                          1553 

In [52]:
dds = DeseqDataSet(
    counts=read_count_dds,
    metadata=sample_file,
    design_factors="Type",
    refit_cooks=True,
    n_cpus=8,
)

In [53]:
dds.deseq2()

Fitting size factors...
... done in 0.03 seconds.

Fitting dispersions...
... done in 1.58 seconds.

Fitting dispersion trend curve...
... done in 2.42 seconds.

Fitting MAP dispersions...
... done in 1.92 seconds.

Fitting LFCs...
... done in 0.93 seconds.

Refitting 80 outliers.

Fitting dispersions...
... done in 0.03 seconds.

Fitting MAP dispersions...
... done in 0.03 seconds.

Fitting LFCs...
... done in 0.03 seconds.



In [54]:
dds

AnnData object with n_obs × n_vars = 59 × 17268
    obs: 'Sex', 'Type'
    uns: 'trend_coeffs', '_squared_logres', 'prior_disp_var'
    obsm: 'design_matrix', 'size_factors', 'replaceable'
    varm: 'non_zero', '_rough_dispersions', 'genewise_dispersions', '_genewise_converged', '_normed_means', 'fitted_dispersions', 'MAP_dispersions', '_MAP_converged', 'dispersions', '_outlier_genes', 'LFC', '_LFC_converged', 'replaced'
    layers: 'normed_counts', '_mu_hat', '_mu_LFC', '_hat_diagonals', 'cooks', 'replace_cooks'

In [55]:
dds.layers["normed_counts"]

array([[  98.60663444,   93.50629128, 1154.37766868, ..., 1913.47874241,
         163.21098114, 3885.61143117],
       [ 117.74463361,  152.9820787 , 1334.72566417, ..., 1387.15210687,
         140.0903305 , 2363.48717093],
       [ 155.31218746,  120.79836803, 1878.2420537 , ..., 1664.94664959,
         171.1885444 , 3029.62307008],
       ...,
       [ 110.4668836 ,    0.        ,  975.15593802, ..., 3609.60065181,
         121.89449225, 7341.09579588],
       [  52.8968589 ,    7.38095706,  837.73862581, ..., 2016.23143569,
         145.15882209, 5260.77714138],
       [  61.2225839 ,   45.21052349, 1095.4133088 , ..., 1264.9527719 ,
         134.68968457, 3722.33310094]])

In [56]:
print(dds.varm["LFC"])

                              intercept  Type_Con_vs_CD  Type_CrF_vs_CD
ENSG00000237491.10_LINC01409   4.912844        0.121062        0.468945
ENSG00000177757.2_FAM87B       4.314602        0.131948       -0.063211
ENSG00000228794.10_LINC01128   6.881496        0.364786        0.009069
ENSG00000187634.13_SAMD11      4.563183        0.608215       -0.164470
ENSG00000188976.11_NOC2L       6.509314       -0.041121       -0.058395
...                                 ...             ...             ...
ENSG00000212907.2_MT-ND4L      6.070070        0.158065       -0.101547
ENSG00000198886.2_MT-ND4       7.410175        0.205518       -0.048156
ENSG00000198786.2_MT-ND5       7.330788        0.293776        0.042844
ENSG00000198695.2_MT-ND6       4.761903        0.297043        0.011883
ENSG00000198727.2_MT-CYB       8.239710        0.159254       -0.043994

[17268 rows x 3 columns]


## Statistical analysis

P-values are computed as earlier from a DeseqStats object with the summary() method, with a new important argument: the contrast. It is a list of three strings of the form ["variable", "tested level", "reference level"] which determines which variable we want to compute LFCs and pvalues for. As an example, to compare the condition B to the condition A, we set contrast=["condition", "B", "A"].

In [57]:
stat_res_Con_vs_CD = DeseqStats(dds, contrast=["Type", "Con", "CD"], n_cpus=8)

- note

If left blank, the variable of interest will be the last one provided in the design_factors attribute of the corresponding DeseqDataSet object, and the reference level will be picked alphabetically. In any case, both variables are still used. This is due to the fact that dds was fit with both as design factors.

In [47]:
stat_res_Con_vs_CD.summary()

Running Wald tests...


Log2 fold change & Wald test p-value: Type Con vs CD


... done in 0.52 seconds.



baseMean  log2FoldChange     lfcSE      stat  \
ENSG00000237491.10_LINC01409   171.584893        0.174655  0.289693  0.602897   
ENSG00000177757.2_FAM87B        75.323918        0.190361  0.412048  0.461988   
ENSG00000228794.10_LINC01128  1071.957048        0.526275  0.174186  3.021345   
ENSG00000187634.13_SAMD11      107.906264        0.877468  0.418874  2.094825   
ENSG00000188976.11_NOC2L       650.471727       -0.059326  0.138463 -0.428460   
...                                   ...             ...       ...       ...   
ENSG00000212907.2_MT-ND4L      432.764866        0.228039  0.225854  1.009679   
ENSG00000198886.2_MT-ND4      1705.550426        0.296500  0.162165  1.828389   
ENSG00000198786.2_MT-ND5      1667.546508        0.423829  0.153746  2.756689   
ENSG00000198695.2_MT-ND6       126.476279        0.428542  0.214457  1.998270   
ENSG00000198727.2_MT-CYB      3868.999654        0.229754  0.187135  1.227750   

                                pvalue      padj  
ENSG00000237491.10_LINC01409  0.546577  0.744454  
ENSG00000177757.2_FAM87B      0.644090  0.811325  
ENSG00000228794.10_LINC01128  0.002517  0.021667  
ENSG00000187634.13_SAMD11     0.036187  0.138629  
ENSG00000188976.11_NOC2L      0.668317  0.826552  
...                                ...       ...  
ENSG00000212907.2_MT-ND4L     0.312649  0.544916  
ENSG00000198886.2_MT-ND4      0.067491  0.210121  
ENSG00000198786.2_MT-ND5      0.005839  0.039560  
ENSG00000198695.2_MT-ND6      0.045687  0.161928  
ENSG00000198727.2_MT-CYB      0.219541  0.442778  

[17268 rows x 6 columns]

In [48]:
stat_res_Con_vs_CrF = DeseqStats(dds, contrast=["Type", "Con", "CrF"], n_cpus=8)
stat_res_Con_vs_CrF.summary()

Running Wald tests...


Log2 fold change & Wald test p-value: Type Con vs CrF


... done in 0.51 seconds.



baseMean  log2FoldChange     lfcSE      stat  \
ENSG00000237491.10_LINC01409   171.584893       -0.501890  0.289404 -1.734219   
ENSG00000177757.2_FAM87B        75.323918        0.281556  0.412430  0.682676   
ENSG00000228794.10_LINC01128  1071.957048        0.513191  0.174232  2.945443   
ENSG00000187634.13_SAMD11      107.906264        1.114748  0.419316  2.658494   
ENSG00000188976.11_NOC2L       650.471727        0.024921  0.138586  0.179820   
...                                   ...             ...       ...       ...   
ENSG00000212907.2_MT-ND4L      432.764866        0.374540  0.225996  1.657283   
ENSG00000198886.2_MT-ND4      1705.550426        0.365975  0.162205  2.256244   
ENSG00000198786.2_MT-ND5      1667.546508        0.362018  0.153772  2.354255   
ENSG00000198695.2_MT-ND6       126.476279        0.411398  0.214762  1.915604   
ENSG00000198727.2_MT-CYB      3868.999654        0.293225  0.187150  1.566794   

                                pvalue      padj  
ENSG00000237491.10_LINC01409  0.082879  0.200558  
ENSG00000177757.2_FAM87B      0.494812  0.674736  
ENSG00000228794.10_LINC01128  0.003225  0.016897  
ENSG00000187634.13_SAMD11     0.007849  0.033721  
ENSG00000188976.11_NOC2L      0.857294  0.920756  
...                                ...       ...  
ENSG00000212907.2_MT-ND4L     0.097462  0.225598  
ENSG00000198886.2_MT-ND4      0.024055  0.079433  
ENSG00000198786.2_MT-ND5      0.018560  0.065222  
ENSG00000198695.2_MT-ND6      0.055416  0.148737  
ENSG00000198727.2_MT-CYB      0.117163  0.256181  

[17268 rows x 6 columns]

In [49]:
stat_res_CD_vs_CrF = DeseqStats(dds, contrast=["Type", "CD", "CrF"], n_cpus=8)
stat_res_CD_vs_CrF.summary()

Running Wald tests...


Log2 fold change & Wald test p-value: Type CD vs CrF


... done in 0.51 seconds.



baseMean  log2FoldChange     lfcSE      stat  \
ENSG00000237491.10_LINC01409   171.584893       -0.676545  0.246323 -2.746573   
ENSG00000177757.2_FAM87B        75.323918        0.091195  0.351241  0.259636   
ENSG00000228794.10_LINC01128  1071.957048       -0.013084  0.148255 -0.088254   
ENSG00000187634.13_SAMD11      107.906264        0.237280  0.357415  0.663878   
ENSG00000188976.11_NOC2L       650.471727        0.084246  0.117925  0.714407   
...                                   ...             ...       ...       ...   
ENSG00000212907.2_MT-ND4L      432.764866        0.146501  0.192314  0.761777   
ENSG00000198886.2_MT-ND4      1705.550426        0.069475  0.137945  0.503643   
ENSG00000198786.2_MT-ND5      1667.546508       -0.061811  0.130801 -0.472556   
ENSG00000198695.2_MT-ND6       126.476279       -0.017144  0.183657 -0.093347   
ENSG00000198727.2_MT-CYB      3868.999654        0.063471  0.159080  0.398985   

                                pvalue      padj  
ENSG00000237491.10_LINC01409  0.006022  0.046633  
ENSG00000177757.2_FAM87B      0.795144  0.896593  
ENSG00000228794.10_LINC01128  0.929674  0.966172  
ENSG00000187634.13_SAMD11     0.506768  0.702846  
ENSG00000188976.11_NOC2L      0.474975  0.676767  
...                                ...       ...  
ENSG00000212907.2_MT-ND4L     0.446193  0.653663  
ENSG00000198886.2_MT-ND4      0.614512  0.782570  
ENSG00000198786.2_MT-ND5      0.636530  0.797384  
ENSG00000198695.2_MT-ND6      0.925628  0.964088  
ENSG00000198727.2_MT-CYB      0.689904  0.831819  

[17268 rows x 6 columns]

#### 문제점

- 비교할 샘플만 뺴서 log2FoldChange와 padj를 계산 해야하는데 다 합쳐서 계산했기 때문에 노이즈가 발생한다. 
- 때문에 비교할 샘플만 뺴는 과정이 필요한데 그 방법을 아직 모르겠음

## LFC shrinkage (multifactor)

In a multifactor setting, LFC shrinkage works as in the single-factor case, but will only shrink the LFCs of a DeseqStats object based on its contrast argument.

In [50]:
stat_res_Con_vs_CD.lfc_shrink(coeff='Type_Con_vs_CD')

Fitting MAP LFCs...


Shrunk log2 fold change & Wald test p-value: Type Con vs CD


... done in 1.95 seconds.



baseMean  log2FoldChange     lfcSE      stat  \
ENSG00000237491.10_LINC01409   171.584893        0.048592  0.155797  0.602897   
ENSG00000177757.2_FAM87B        75.323918        0.029496  0.163841  0.461988   
ENSG00000228794.10_LINC01128  1071.957048        0.426224  0.183985  3.021345   
ENSG00000187634.13_SAMD11      107.906264        0.210879  0.330057  2.094825   
ENSG00000188976.11_NOC2L       650.471727       -0.036396  0.110127 -0.428460   
...                                   ...             ...       ...       ...   
ENSG00000212907.2_MT-ND4L      432.764866        0.093984  0.155462  1.009679   
ENSG00000198886.2_MT-ND4      1705.550426        0.201255  0.152817  1.828389   
ENSG00000198786.2_MT-ND5      1667.546508        0.346431  0.159429  2.756689   
ENSG00000198695.2_MT-ND6       126.476279        0.246385  0.210898  1.998270   
ENSG00000198727.2_MT-CYB      3868.999654        0.141726  0.155497  1.227750   

                                pvalue      padj  
ENSG00000237491.10_LINC01409  0.546577  0.744454  
ENSG00000177757.2_FAM87B      0.644090  0.811325  
ENSG00000228794.10_LINC01128  0.002517  0.021667  
ENSG00000187634.13_SAMD11     0.036187  0.138629  
ENSG00000188976.11_NOC2L      0.668317  0.826552  
...                                ...       ...  
ENSG00000212907.2_MT-ND4L     0.312649  0.544916  
ENSG00000198886.2_MT-ND4      0.067491  0.210121  
ENSG00000198786.2_MT-ND5      0.005839  0.039560  
ENSG00000198695.2_MT-ND6      0.045687  0.161928  
ENSG00000198727.2_MT-CYB      0.219541  0.442778  

[17268 rows x 6 columns]

In [51]:
stat_res_Con_vs_CrF.lfc_shrink(coeff='Type_Con_vs_CrF')

KeyError: "The coeff argument 'Type_Con_vs_CrF' should be one the LFC columns. The available LFC coeffs are Index(['Type_Con_vs_CD', 'Type_CrF_vs_CD'], dtype='object'). If the desired coefficient is not available, please set `ref_level = [Type, CrF]` in DeseqDataSet and rerun."